internal/controllers/machinedeployment/mdutil/util.go

```diff
@@ -184,59 +183,59 @@ func getIntFromAnnotation(ms *clusterv1.MachineSet, annotationKey string, logger
 	return int32(intValue), true
 }
 
-// SetNewMachineSetAnnotations sets new machine set's annotations appropriately by updating its revision and
-// copying required deployment annotations to it; it returns true if machine set's annotation is changed.
-func SetNewMachineSetAnnotations(deployment *clusterv1.MachineDeployment, newMS *clusterv1.MachineSet, newRevision string, exists bool, logger logr.Logger) bool {
-	logger = logger.WithValues("MachineSet", klog.KObj(newMS))
-
-	// First, copy deployment's annotations (except for apply and revision annotations)
-	annotationChanged := copyDeploymentAnnotationsToMachineSet(deployment, newMS)
-	// Then, update machine set's revision annotation
-	if newMS.Annotations == nil {
-		newMS.Annotations = make(map[string]string)
+// ComputeMachineSetAnnotations computes the annotations that should be set on the MachineSet.
+// Note: The passed in newMS is nil if the new MachineSet doesn't exist in the apiserver yet.
+func ComputeMachineSetAnnotations(log logr.Logger, deployment *clusterv1.MachineDeployment, oldMSs []*clusterv1.MachineSet, newMS *clusterv1.MachineSet) (map[string]string, error) {
```

Comment:
Please add a unit test checking newMS nil, newMS != nill and all the weirdness of revision history